{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dcb26a19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jalh/anaconda3/envs/gaishi-analysis150126/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.7.3' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from sprime_mapping import map_sprime_segments\n",
    "from sprime_info import compute_sprime_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "165acd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# these two files are necessary: score_file with SPrime output and original vcf_file output\n",
    "vcf_file = \"archie.1_biallelic.vcf.gz\"\n",
    "score_file = \"sprime.1src.out.100000.score\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4766a6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for getting only the introgressed tracts in the target individuals, we also have to provide a list with the individual names (have to correspond to the vcf-file) we are interested in\n",
    "# if not provided, for all individuals in the vcf-file the introgression status is checked (including reference individuals etc. - this is usually not desired / reasonable, but perhaps one want to check)\n",
    "tgt_individuals_file = \"archie.1.tgt.ind.list\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "709e4d40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if we ant to write the file, also an out_file has to be scpecified\n",
    "out_file_wo_filtering = \"archie.0.sprime.100000.inferred_wo_filtering.bed\"\n",
    "out_file_every_snp = \"archie.0.sprime.100000.inferred_every_snp.bed\"\n",
    "out_file_frac_50 = \"archie.0.sprime.100000.inferred_fraction_050.bed\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "15c4179f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open file with tgt individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a5c5620",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(tgt_individuals_file) as f:\n",
    "    tgt_individuals = [line.strip() for line in f if line.strip()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7575004e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['tsk_50',\n",
       " 'tsk_51',\n",
       " 'tsk_52',\n",
       " 'tsk_53',\n",
       " 'tsk_54',\n",
       " 'tsk_55',\n",
       " 'tsk_56',\n",
       " 'tsk_57',\n",
       " 'tsk_58',\n",
       " 'tsk_59']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tgt_individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "29339831",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's also check the sprime output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c0cdbb4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    SEGMENT  SNP_COUNT\n",
      "0         0         75\n",
      "1         1         47\n",
      "2         4         42\n",
      "3         5         36\n",
      "4         2         31\n",
      "5         3         30\n",
      "6        11         29\n",
      "7        10         28\n",
      "8         6         27\n",
      "9         7         24\n",
      "10        9         23\n",
      "11        8         22\n",
      "12       13         21\n",
      "13       12         20\n",
      "455\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(score_file, sep=\"\\t\")\n",
    "\n",
    "# Count the number of SNPs per SEGMENT\n",
    "snp_counts = df['SEGMENT'].value_counts()\n",
    "\n",
    "snp_counts_df = snp_counts.reset_index()\n",
    "snp_counts_df.columns = ['SEGMENT', 'SNP_COUNT']\n",
    "\n",
    "print(snp_counts_df)\n",
    "print(snp_counts_df[\"SNP_COUNT\"].sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57954fbd",
   "metadata": {},
   "source": [
    "### In this run of the function, we create an output-file (bed-file)\n",
    "## We do not apply any filtering (see below for possible filtering steps)\n",
    "## With these settings, the first and last SNP, which matches the archaic SNPs of the segment, is used as border of the segment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "49402de7",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    out_file_wo_filtering, only_tract_output=True, phased=True, target_individuals=tgt_individuals, return_full_records=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ace67021",
   "metadata": {},
   "source": [
    "### we can also show the full records, including the nsnps in each segment, the fraction of archaic snps in the segment for a specific individual and the snps present\n",
    "### out_file is from now set to None, so no new out_file is created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "69ed2d18",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    None, only_tract_output=True, phased=True, target_individuals=tgt_individuals, return_full_records=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5a3c9b5a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>all_snps_present</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>267892</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>279660</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>267891</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>0.785714</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[267891, 268554, 279659, 281915, 290557, 29327...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end individual  haplotype  nsnps  segment_fraction  \\\n",
       "0      1  238504  314341     tsk_50          1     28          1.000000   \n",
       "1      1  238504  267892     tsk_52          1      7          0.250000   \n",
       "2      1  238504  279660     tsk_53          1      9          0.321429   \n",
       "3      1  238504  314341     tsk_57          1     28          1.000000   \n",
       "4      1  267891  314341     tsk_58          2     22          0.785714   \n",
       "\n",
       "   segment                                   all_snps_present  \\\n",
       "0       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "1       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "2       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "3       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "4       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "\n",
       "                                archaic_snps_present  \n",
       "0  [238504, 251513, 255766, 257368, 257723, 25802...  \n",
       "1  [238504, 251513, 255766, 257368, 257723, 25802...  \n",
       "2  [238504, 251513, 255766, 257368, 257723, 25802...  \n",
       "3  [238504, 251513, 255766, 257368, 257723, 25802...  \n",
       "4  [267891, 268554, 279659, 281915, 290557, 29327...  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6d654775",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>all_snps_present</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>1</td>\n",
       "      <td>21560934</td>\n",
       "      <td>21560935</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>1</td>\n",
       "      <td>[21541283, 21541302, 21542755, 21544262, 21544...</td>\n",
       "      <td>[21560934]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>1</td>\n",
       "      <td>21560934</td>\n",
       "      <td>21560935</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>1</td>\n",
       "      <td>[21541283, 21541302, 21542755, 21544262, 21544...</td>\n",
       "      <td>[21560934]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1</td>\n",
       "      <td>21560934</td>\n",
       "      <td>21560935</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>1</td>\n",
       "      <td>[21541283, 21541302, 21542755, 21544262, 21544...</td>\n",
       "      <td>[21560934]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>1</td>\n",
       "      <td>21905012</td>\n",
       "      <td>22041832</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>3</td>\n",
       "      <td>[21905012, 21906798, 21911405, 21915067, 21916...</td>\n",
       "      <td>[21905012, 21906798, 21911405, 21915067, 21916...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>1</td>\n",
       "      <td>21927036</td>\n",
       "      <td>22028446</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>0.200000</td>\n",
       "      <td>3</td>\n",
       "      <td>[21905012, 21906798, 21911405, 21915067, 21916...</td>\n",
       "      <td>[21927036, 21932237, 21958351, 21991539, 22000...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end individual  haplotype  nsnps  segment_fraction  \\\n",
       "52      1  21560934  21560935     tsk_57          2      1          0.021277   \n",
       "53      1  21560934  21560935     tsk_59          1      1          0.021277   \n",
       "54      1  21560934  21560935     tsk_59          2      1          0.021277   \n",
       "55      1  21905012  22041832     tsk_56          1     24          0.800000   \n",
       "56      1  21927036  22028446     tsk_56          2      6          0.200000   \n",
       "\n",
       "    segment                                   all_snps_present  \\\n",
       "52        1  [21541283, 21541302, 21542755, 21544262, 21544...   \n",
       "53        1  [21541283, 21541302, 21542755, 21544262, 21544...   \n",
       "54        1  [21541283, 21541302, 21542755, 21544262, 21544...   \n",
       "55        3  [21905012, 21906798, 21911405, 21915067, 21916...   \n",
       "56        3  [21905012, 21906798, 21911405, 21915067, 21916...   \n",
       "\n",
       "                                 archaic_snps_present  \n",
       "52                                         [21560934]  \n",
       "53                                         [21560934]  \n",
       "54                                         [21560934]  \n",
       "55  [21905012, 21906798, 21911405, 21915067, 21916...  \n",
       "56  [21927036, 21932237, 21958351, 21991539, 22000...  "
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5240c18b",
   "metadata": {},
   "source": [
    "# Computation of averages\n",
    "\n",
    "### We now additionally compute the length of the regions as well as the segments per individual\n",
    "## afterwards, we can average over these values and get the desired results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "111af707",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_wo_filtering = compute_sprime_info(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "40fc4de7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>all_snps_present</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>interval_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>75837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>267892</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>tsk_52_1</td>\n",
       "      <td>29388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>279660</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>0.321429</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>41156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>75837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>267891</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "      <td>22</td>\n",
       "      <td>0.785714</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[267891, 268554, 279659, 281915, 290557, 29327...</td>\n",
       "      <td>tsk_58_2</td>\n",
       "      <td>46450</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom   start     end individual  haplotype  nsnps  segment_fraction  \\\n",
       "0      1  238504  314341     tsk_50          1     28          1.000000   \n",
       "1      1  238504  267892     tsk_52          1      7          0.250000   \n",
       "2      1  238504  279660     tsk_53          1      9          0.321429   \n",
       "3      1  238504  314341     tsk_57          1     28          1.000000   \n",
       "4      1  267891  314341     tsk_58          2     22          0.785714   \n",
       "\n",
       "   segment                                   all_snps_present  \\\n",
       "0       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "1       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "2       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "3       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "4       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "\n",
       "                                archaic_snps_present individual_haplotype  \\\n",
       "0  [238504, 251513, 255766, 257368, 257723, 25802...             tsk_50_1   \n",
       "1  [238504, 251513, 255766, 257368, 257723, 25802...             tsk_52_1   \n",
       "2  [238504, 251513, 255766, 257368, 257723, 25802...             tsk_53_1   \n",
       "3  [238504, 251513, 255766, 257368, 257723, 25802...             tsk_57_1   \n",
       "4  [267891, 268554, 279659, 281915, 290557, 29327...             tsk_58_2   \n",
       "\n",
       "   interval_length  \n",
       "0            75837  \n",
       "1            29388  \n",
       "2            41156  \n",
       "3            75837  \n",
       "4            46450  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the df now also has interval length\n",
    "res_wo_filtering[\"df\"].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6071d470",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>total_length</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>192915</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>118567</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>45941</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_51_2</td>\n",
       "      <td>92132</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_52_1</td>\n",
       "      <td>29389</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>199510</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>178030</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>178958</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_54_1</td>\n",
       "      <td>13709</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>107553</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tsk_55_1</td>\n",
       "      <td>33357</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>107553</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tsk_56_1</td>\n",
       "      <td>420797</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>172496</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>448502</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tsk_57_2</td>\n",
       "      <td>65310</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>311758</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tsk_58_2</td>\n",
       "      <td>320244</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>325344</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tsk_59_2</td>\n",
       "      <td>255238</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual_haplotype  total_length individual  haplotype\n",
       "0              tsk_50_1        192915     tsk_50          1\n",
       "1              tsk_50_2        118567     tsk_50          2\n",
       "2              tsk_51_1         45941     tsk_51          1\n",
       "3              tsk_51_2         92132     tsk_51          2\n",
       "4              tsk_52_1         29389     tsk_52          1\n",
       "5              tsk_52_2        199510     tsk_52          2\n",
       "6              tsk_53_1        178030     tsk_53          1\n",
       "7              tsk_53_2        178958     tsk_53          2\n",
       "8              tsk_54_1         13709     tsk_54          1\n",
       "9              tsk_54_2        107553     tsk_54          2\n",
       "10             tsk_55_1         33357     tsk_55          1\n",
       "11             tsk_55_2        107553     tsk_55          2\n",
       "12             tsk_56_1        420797     tsk_56          1\n",
       "13             tsk_56_2        172496     tsk_56          2\n",
       "14             tsk_57_1        448502     tsk_57          1\n",
       "15             tsk_57_2         65310     tsk_57          2\n",
       "16             tsk_58_1        311758     tsk_58          1\n",
       "17             tsk_58_2        320244     tsk_58          2\n",
       "18             tsk_59_1        325344     tsk_59          1\n",
       "19             tsk_59_2        255238     tsk_59          2"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total length of introgressed regions per individual/haplotype\n",
    "res_wo_filtering[\"total_length_ind_hap\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "92a4cba3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>segment</th>\n",
       "      <th>total_length</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>1</td>\n",
       "      <td>117078</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>10</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>1</td>\n",
       "      <td>72626</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>13</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual_haplotype  segment  total_length individual  haplotype\n",
       "0             tsk_50_1        1        117078     tsk_50          1\n",
       "1             tsk_50_1       10         75837     tsk_50          1\n",
       "2             tsk_50_2        0             1     tsk_50          2\n",
       "3             tsk_50_2        1         72626     tsk_50          2\n",
       "4             tsk_50_2       13         45940     tsk_50          2"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total length of introgressed regions per individual/haplotype and segment\n",
    "res_wo_filtering[\"total_length_ind_hap_seg\"].iloc[0:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "b843105c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>segments</th>\n",
       "      <th>nr_segments</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>[1, 10]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>[0, 1, 13]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>[1, 13]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_51_2</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_52_1</td>\n",
       "      <td>[1, 10]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>[1, 4, 5]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>[0, 1, 6, 10]</td>\n",
       "      <td>4</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>[1, 4]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_54_1</td>\n",
       "      <td>[1, 4]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>[1, 6]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tsk_55_1</td>\n",
       "      <td>[1, 4]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>[1, 6]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tsk_56_1</td>\n",
       "      <td>[0, 1, 3]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>[0, 1, 3, 11]</td>\n",
       "      <td>4</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>[0, 1, 7, 10]</td>\n",
       "      <td>4</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tsk_57_2</td>\n",
       "      <td>[1, 4, 7]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>[1, 8, 12]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tsk_58_2</td>\n",
       "      <td>[1, 8, 10, 12]</td>\n",
       "      <td>4</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>[1, 2, 4, 5, 9]</td>\n",
       "      <td>5</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tsk_59_2</td>\n",
       "      <td>[1, 2, 9]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual_haplotype         segments  nr_segments individual  haplotype\n",
       "0              tsk_50_1          [1, 10]            2     tsk_50          1\n",
       "1              tsk_50_2       [0, 1, 13]            3     tsk_50          2\n",
       "2              tsk_51_1          [1, 13]            2     tsk_51          1\n",
       "3              tsk_51_2           [0, 1]            2     tsk_51          2\n",
       "4              tsk_52_1          [1, 10]            2     tsk_52          1\n",
       "5              tsk_52_2        [1, 4, 5]            3     tsk_52          2\n",
       "6              tsk_53_1    [0, 1, 6, 10]            4     tsk_53          1\n",
       "7              tsk_53_2           [1, 4]            2     tsk_53          2\n",
       "8              tsk_54_1           [1, 4]            2     tsk_54          1\n",
       "9              tsk_54_2           [1, 6]            2     tsk_54          2\n",
       "10             tsk_55_1           [1, 4]            2     tsk_55          1\n",
       "11             tsk_55_2           [1, 6]            2     tsk_55          2\n",
       "12             tsk_56_1        [0, 1, 3]            3     tsk_56          1\n",
       "13             tsk_56_2    [0, 1, 3, 11]            4     tsk_56          2\n",
       "14             tsk_57_1    [0, 1, 7, 10]            4     tsk_57          1\n",
       "15             tsk_57_2        [1, 4, 7]            3     tsk_57          2\n",
       "16             tsk_58_1       [1, 8, 12]            3     tsk_58          1\n",
       "17             tsk_58_2   [1, 8, 10, 12]            4     tsk_58          2\n",
       "18             tsk_59_1  [1, 2, 4, 5, 9]            5     tsk_59          1\n",
       "19             tsk_59_2        [1, 2, 9]            3     tsk_59          2"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#segments per individual/haplotype \n",
    "res_wo_filtering[\"segments_per_ind\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aaeaba68",
   "metadata": {},
   "source": [
    "## Average number of segments per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "aa382008",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_per_individual = (\n",
    "    res_wo_filtering[\"segments_per_ind\"].groupby(\"individual\")[\"nr_segments\"]\n",
    "      .mean()\n",
    "      .reset_index(name=\"avg_nr_segments\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "56cc0dc3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>avg_nr_segments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_51</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_53</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_56</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_57</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_58</td>\n",
       "      <td>3.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_59</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  avg_nr_segments\n",
       "0     tsk_50              2.5\n",
       "1     tsk_51              2.0\n",
       "2     tsk_52              2.5\n",
       "3     tsk_53              3.0\n",
       "4     tsk_54              2.0\n",
       "5     tsk_55              2.0\n",
       "6     tsk_56              3.5\n",
       "7     tsk_57              3.5\n",
       "8     tsk_58              3.5\n",
       "9     tsk_59              4.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_per_individual"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d532f57d",
   "metadata": {},
   "source": [
    "## or per chromosome, ,i.e. individual haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b6d0e048",
   "metadata": {},
   "outputs": [],
   "source": [
    "avg_per_individual_and_haplotype = (\n",
    "    res_wo_filtering[\"segments_per_ind\"].groupby([\"individual\", \"haplotype\"])[\"nr_segments\"]\n",
    "      .mean()\n",
    "      .reset_index(name=\"avg_nr_segments\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b2744cc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>avg_nr_segments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_51</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_52</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_54</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tsk_55</td>\n",
       "      <td>1</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tsk_57</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual  haplotype  avg_nr_segments\n",
       "0      tsk_50          1              2.0\n",
       "1      tsk_50          2              3.0\n",
       "2      tsk_51          1              2.0\n",
       "3      tsk_51          2              2.0\n",
       "4      tsk_52          1              2.0\n",
       "5      tsk_52          2              3.0\n",
       "6      tsk_53          1              4.0\n",
       "7      tsk_53          2              2.0\n",
       "8      tsk_54          1              2.0\n",
       "9      tsk_54          2              2.0\n",
       "10     tsk_55          1              2.0\n",
       "11     tsk_55          2              2.0\n",
       "12     tsk_56          1              3.0\n",
       "13     tsk_56          2              4.0\n",
       "14     tsk_57          1              4.0\n",
       "15     tsk_57          2              3.0\n",
       "16     tsk_58          1              3.0\n",
       "17     tsk_58          2              4.0\n",
       "18     tsk_59          1              5.0\n",
       "19     tsk_59          2              3.0"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "avg_per_individual_and_haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a8431016",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Filtering\n",
    "\n",
    "### we can specify minimum fractions for the amount of archaic snps which has to be present in an individual\n",
    "### now get fewer hits (in particular, one can also set a threshold afterwards, because the fraction is stored in segment_faction)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "b8ed47f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_50 = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    out_file_frac_50, only_tract_output=True, phased=True, return_full_records=False, target_individuals=tgt_individuals,\n",
    "    segment_fraction=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "b4952374",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_50_full = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    None, only_tract_output=True, phased=True, return_full_records=True, target_individuals=tgt_individuals,\n",
    "    segment_fraction=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "7b0bd189",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n"
     ]
    }
   ],
   "source": [
    "# now the number of fragments in the dataframe is considerably smaller \n",
    "print(len(df_50_full))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "57c4e7c2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>all_snps_present</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1</td>\n",
       "      <td>20647984</td>\n",
       "      <td>20693924</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>13</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1</td>\n",
       "      <td>20647984</td>\n",
       "      <td>20693924</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>13</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1</td>\n",
       "      <td>21541283</td>\n",
       "      <td>21572805</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>0.574468</td>\n",
       "      <td>1</td>\n",
       "      <td>[21541283, 21541302, 21542755, 21544262, 21544...</td>\n",
       "      <td>[21541283, 21541302, 21542755, 21544262, 21544...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1</td>\n",
       "      <td>21557378</td>\n",
       "      <td>21674456</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>33</td>\n",
       "      <td>0.702128</td>\n",
       "      <td>1</td>\n",
       "      <td>[21541283, 21541302, 21542755, 21544262, 21544...</td>\n",
       "      <td>[21557378, 21559072, 21560497, 21560541, 21560...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1</td>\n",
       "      <td>21905012</td>\n",
       "      <td>22041832</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>24</td>\n",
       "      <td>0.800000</td>\n",
       "      <td>3</td>\n",
       "      <td>[21905012, 21906798, 21911405, 21915067, 21916...</td>\n",
       "      <td>[21905012, 21906798, 21911405, 21915067, 21916...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end individual  haplotype  nsnps  segment_fraction  \\\n",
       "19      1  20647984  20693924     tsk_50          2     21          1.000000   \n",
       "20      1  20647984  20693924     tsk_51          1     21          1.000000   \n",
       "21      1  21541283  21572805     tsk_58          1     27          0.574468   \n",
       "22      1  21557378  21674456     tsk_50          1     33          0.702128   \n",
       "23      1  21905012  22041832     tsk_56          1     24          0.800000   \n",
       "\n",
       "    segment                                   all_snps_present  \\\n",
       "19       13  [20647984, 20648821, 20649517, 20649859, 20650...   \n",
       "20       13  [20647984, 20648821, 20649517, 20649859, 20650...   \n",
       "21        1  [21541283, 21541302, 21542755, 21544262, 21544...   \n",
       "22        1  [21541283, 21541302, 21542755, 21544262, 21544...   \n",
       "23        3  [21905012, 21906798, 21911405, 21915067, 21916...   \n",
       "\n",
       "                                 archaic_snps_present  \n",
       "19  [20647984, 20648821, 20649517, 20649859, 20650...  \n",
       "20  [20647984, 20648821, 20649517, 20649859, 20650...  \n",
       "21  [21541283, 21541302, 21542755, 21544262, 21544...  \n",
       "22  [21557378, 21559072, 21560497, 21560541, 21560...  \n",
       "23  [21905012, 21906798, 21911405, 21915067, 21916...  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_50_full.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "9c8d3b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_frac50 = compute_sprime_info(df_50_full)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "1e3ed8ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>total_length</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>192915</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>199509</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>178957</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_56_1</td>\n",
       "      <td>420796</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>71084</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>448501</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>311758</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tsk_58_2</td>\n",
       "      <td>46450</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>325342</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tsk_59_2</td>\n",
       "      <td>123895</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual_haplotype  total_length individual  haplotype\n",
       "0              tsk_50_1        192915     tsk_50          1\n",
       "1              tsk_50_2         45940     tsk_50          2\n",
       "2              tsk_51_1         45940     tsk_51          1\n",
       "3              tsk_52_2        199509     tsk_52          2\n",
       "4              tsk_53_1        107552     tsk_53          1\n",
       "5              tsk_53_2        178957     tsk_53          2\n",
       "6              tsk_54_2        107552     tsk_54          2\n",
       "7              tsk_55_2        107552     tsk_55          2\n",
       "8              tsk_56_1        420796     tsk_56          1\n",
       "9              tsk_56_2         71084     tsk_56          2\n",
       "10             tsk_57_1        448501     tsk_57          1\n",
       "11             tsk_58_1        311758     tsk_58          1\n",
       "12             tsk_58_2         46450     tsk_58          2\n",
       "13             tsk_59_1        325342     tsk_59          1\n",
       "14             tsk_59_2        123895     tsk_59          2"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total length of introgressed regions per individual/haplotype\n",
    "res_frac50[\"total_length_ind_hap\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b7f2be08",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>segment</th>\n",
       "      <th>total_length</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>1</td>\n",
       "      <td>117078</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>10</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>13</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>13</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>4</td>\n",
       "      <td>141874</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>5</td>\n",
       "      <td>57635</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>6</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>4</td>\n",
       "      <td>178957</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>6</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>6</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tsk_56_1</td>\n",
       "      <td>0</td>\n",
       "      <td>283976</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tsk_56_1</td>\n",
       "      <td>3</td>\n",
       "      <td>136820</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>11</td>\n",
       "      <td>71084</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>0</td>\n",
       "      <td>283976</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>7</td>\n",
       "      <td>88688</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>10</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>1</td>\n",
       "      <td>31522</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>8</td>\n",
       "      <td>193764</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>12</td>\n",
       "      <td>86472</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>tsk_58_2</td>\n",
       "      <td>10</td>\n",
       "      <td>46450</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>2</td>\n",
       "      <td>167202</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>4</td>\n",
       "      <td>117664</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>5</td>\n",
       "      <td>40476</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>tsk_59_2</td>\n",
       "      <td>9</td>\n",
       "      <td>123895</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual_haplotype  segment  total_length individual  haplotype\n",
       "0              tsk_50_1        1        117078     tsk_50          1\n",
       "1              tsk_50_1       10         75837     tsk_50          1\n",
       "2              tsk_50_2       13         45940     tsk_50          2\n",
       "3              tsk_51_1       13         45940     tsk_51          1\n",
       "4              tsk_52_2        4        141874     tsk_52          2\n",
       "5              tsk_52_2        5         57635     tsk_52          2\n",
       "6              tsk_53_1        6        107552     tsk_53          1\n",
       "7              tsk_53_2        4        178957     tsk_53          2\n",
       "8              tsk_54_2        6        107552     tsk_54          2\n",
       "9              tsk_55_2        6        107552     tsk_55          2\n",
       "10             tsk_56_1        0        283976     tsk_56          1\n",
       "11             tsk_56_1        3        136820     tsk_56          1\n",
       "12             tsk_56_2       11         71084     tsk_56          2\n",
       "13             tsk_57_1        0        283976     tsk_57          1\n",
       "14             tsk_57_1        7         88688     tsk_57          1\n",
       "15             tsk_57_1       10         75837     tsk_57          1\n",
       "16             tsk_58_1        1         31522     tsk_58          1\n",
       "17             tsk_58_1        8        193764     tsk_58          1\n",
       "18             tsk_58_1       12         86472     tsk_58          1\n",
       "19             tsk_58_2       10         46450     tsk_58          2\n",
       "20             tsk_59_1        2        167202     tsk_59          1\n",
       "21             tsk_59_1        4        117664     tsk_59          1\n",
       "22             tsk_59_1        5         40476     tsk_59          1\n",
       "23             tsk_59_2        9        123895     tsk_59          2"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total length of introgressed regions per individual/haplotype and segment\n",
    "res_frac50[\"total_length_ind_hap_seg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c1402274",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>segments</th>\n",
       "      <th>nr_segments</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>[1, 10]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>[13]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>[13]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>[4, 5]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>[6]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>[4]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>[6]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>[6]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_56_1</td>\n",
       "      <td>[0, 3]</td>\n",
       "      <td>2</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>[0, 7, 10]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>tsk_58_1</td>\n",
       "      <td>[1, 8, 12]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>tsk_58_2</td>\n",
       "      <td>[10]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_58</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>tsk_59_1</td>\n",
       "      <td>[2, 4, 5]</td>\n",
       "      <td>3</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>tsk_59_2</td>\n",
       "      <td>[9]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_59</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   individual_haplotype    segments  nr_segments individual  haplotype\n",
       "0              tsk_50_1     [1, 10]            2     tsk_50          1\n",
       "1              tsk_50_2        [13]            1     tsk_50          2\n",
       "2              tsk_51_1        [13]            1     tsk_51          1\n",
       "3              tsk_52_2      [4, 5]            2     tsk_52          2\n",
       "4              tsk_53_1         [6]            1     tsk_53          1\n",
       "5              tsk_53_2         [4]            1     tsk_53          2\n",
       "6              tsk_54_2         [6]            1     tsk_54          2\n",
       "7              tsk_55_2         [6]            1     tsk_55          2\n",
       "8              tsk_56_1      [0, 3]            2     tsk_56          1\n",
       "9              tsk_56_2        [11]            1     tsk_56          2\n",
       "10             tsk_57_1  [0, 7, 10]            3     tsk_57          1\n",
       "11             tsk_58_1  [1, 8, 12]            3     tsk_58          1\n",
       "12             tsk_58_2        [10]            1     tsk_58          2\n",
       "13             tsk_59_1   [2, 4, 5]            3     tsk_59          1\n",
       "14             tsk_59_2         [9]            1     tsk_59          2"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#segments per individual/haplotype \n",
    "res_frac50[\"segments_per_ind\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b722bd8",
   "metadata": {},
   "source": [
    "### if we addionally specify min_snps, we get more fragments, because within the segment, fragments of contiguous archaic SNPs with a length of at least 3 are determined and these fragments are added to the final list (whereas shorter fragments are discarded)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "7edf969f",
   "metadata": {},
   "outputs": [],
   "source": [
    "dff_frac50_minsnps3 = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    None, only_tract_output=True, phased=True, return_full_records=True, target_individuals=tgt_individuals,\n",
    "    segment_fraction=0.5, min_snps=3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "66c65362",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "37\n"
     ]
    }
   ],
   "source": [
    "# now the number of fragments in the dataframe is considerably smaller \n",
    "print(len(dff_frac50_minsnps3))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "b16d9686",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1</td>\n",
       "      <td>21905012</td>\n",
       "      <td>21923919</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>0.8</td>\n",
       "      <td>3</td>\n",
       "      <td>[21905012, 21906798, 21911405, 21915067, 21916...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1</td>\n",
       "      <td>21935944</td>\n",
       "      <td>21951411</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.8</td>\n",
       "      <td>3</td>\n",
       "      <td>[21935944, 21938125, 21949125, 21950785, 21951...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1</td>\n",
       "      <td>21964563</td>\n",
       "      <td>21978154</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0.8</td>\n",
       "      <td>3</td>\n",
       "      <td>[21964563, 21966130, 21970779, 21977944, 21978...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1</td>\n",
       "      <td>21994331</td>\n",
       "      <td>21998049</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>0.8</td>\n",
       "      <td>3</td>\n",
       "      <td>[21994331, 21995907, 21998048]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1</td>\n",
       "      <td>22033401</td>\n",
       "      <td>22041832</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.8</td>\n",
       "      <td>3</td>\n",
       "      <td>[22033401, 22036294, 22038925, 22041831]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom     start       end individual  haplotype  nsnps  segment_fraction  \\\n",
       "32      1  21905012  21923919     tsk_56          1      6               0.8   \n",
       "33      1  21935944  21951411     tsk_56          1      5               0.8   \n",
       "34      1  21964563  21978154     tsk_56          1      5               0.8   \n",
       "35      1  21994331  21998049     tsk_56          1      3               0.8   \n",
       "36      1  22033401  22041832     tsk_56          1      4               0.8   \n",
       "\n",
       "    segment                               archaic_snps_present  \n",
       "32        3  [21905012, 21906798, 21911405, 21915067, 21916...  \n",
       "33        3  [21935944, 21938125, 21949125, 21950785, 21951...  \n",
       "34        3  [21964563, 21966130, 21970779, 21977944, 21978...  \n",
       "35        3                     [21994331, 21995907, 21998048]  \n",
       "36        3           [22033401, 22036294, 22038925, 22041831]  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dff_frac50_minsnps3.tail()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "44219d6a",
   "metadata": {},
   "source": [
    "### Finally, by setting single_snp_bed=True, we can map every single SNPs (so also in the bed-file there are exclusively entries of length 1, one entry for each SNP, no region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "afddb067",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_single_snp = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    out_file_every_snp, only_tract_output=True, phased=True, return_full_records=False, target_individuals=tgt_individuals,\n",
    "    single_snp_bed=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bdeffb6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_single_snp = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    None, only_tract_output=True, phased=True, return_full_records=True, target_individuals=tgt_individuals,\n",
    "    single_snp_bed=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "0bb778ce",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>238505</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>251513</td>\n",
       "      <td>251514</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>10</td>\n",
       "      <td>[251513]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>255766</td>\n",
       "      <td>255767</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>10</td>\n",
       "      <td>[255766]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>257368</td>\n",
       "      <td>257369</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>10</td>\n",
       "      <td>[257368]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>257723</td>\n",
       "      <td>257724</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0.035714</td>\n",
       "      <td>10</td>\n",
       "      <td>[257723]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>866</th>\n",
       "      <td>1</td>\n",
       "      <td>21932237</td>\n",
       "      <td>21932238</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>3</td>\n",
       "      <td>[21932237]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>867</th>\n",
       "      <td>1</td>\n",
       "      <td>21958351</td>\n",
       "      <td>21958352</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>3</td>\n",
       "      <td>[21958351]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>868</th>\n",
       "      <td>1</td>\n",
       "      <td>21991539</td>\n",
       "      <td>21991540</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>3</td>\n",
       "      <td>[21991539]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>869</th>\n",
       "      <td>1</td>\n",
       "      <td>22000748</td>\n",
       "      <td>22000749</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>3</td>\n",
       "      <td>[22000748]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>870</th>\n",
       "      <td>1</td>\n",
       "      <td>22028445</td>\n",
       "      <td>22028446</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0.033333</td>\n",
       "      <td>3</td>\n",
       "      <td>[22028445]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>871 rows  9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     chrom     start       end individual  haplotype  nsnps  segment_fraction  \\\n",
       "0        1    238504    238505     tsk_50          1      1          0.035714   \n",
       "1        1    251513    251514     tsk_50          1      1          0.035714   \n",
       "2        1    255766    255767     tsk_50          1      1          0.035714   \n",
       "3        1    257368    257369     tsk_50          1      1          0.035714   \n",
       "4        1    257723    257724     tsk_50          1      1          0.035714   \n",
       "..     ...       ...       ...        ...        ...    ...               ...   \n",
       "866      1  21932237  21932238     tsk_56          2      1          0.033333   \n",
       "867      1  21958351  21958352     tsk_56          2      1          0.033333   \n",
       "868      1  21991539  21991540     tsk_56          2      1          0.033333   \n",
       "869      1  22000748  22000749     tsk_56          2      1          0.033333   \n",
       "870      1  22028445  22028446     tsk_56          2      1          0.033333   \n",
       "\n",
       "     segment archaic_snps_present  \n",
       "0         10             [238504]  \n",
       "1         10             [251513]  \n",
       "2         10             [255766]  \n",
       "3         10             [257368]  \n",
       "4         10             [257723]  \n",
       "..       ...                  ...  \n",
       "866        3           [21932237]  \n",
       "867        3           [21958351]  \n",
       "868        3           [21991539]  \n",
       "869        3           [22000748]  \n",
       "870        3           [22028445]  \n",
       "\n",
       "[871 rows x 9 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_single_snp"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1091ee5f",
   "metadata": {},
   "source": [
    "### in this case, the function gives us directly the number of archaic SNPs per individual/haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a7f7bb73",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "individual\n",
      "tsk_57    140\n",
      "tsk_56    136\n",
      "tsk_59    110\n",
      "tsk_58     92\n",
      "tsk_53     86\n",
      "tsk_50     85\n",
      "tsk_52     83\n",
      "tsk_51     57\n",
      "tsk_54     43\n",
      "tsk_55     39\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "archaic_snp_counts_per_individual = df_single_snp['individual'].value_counts()\n",
    "print(archaic_snp_counts_per_individual)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10fbf4eb",
   "metadata": {},
   "source": [
    "### and we can also easy calcualte the avgerage archaic SNPs per individual\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "782ad984",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "87.1"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "archaic_snp_counts_per_individual.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "b98361b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "individual  haplotype\n",
      "tsk_57      1            126\n",
      "tsk_56      1             99\n",
      "tsk_59      1             75\n",
      "tsk_52      2             75\n",
      "tsk_50      1             61\n",
      "tsk_58      1             50\n",
      "tsk_53      1             43\n",
      "            2             43\n",
      "tsk_58      2             42\n",
      "tsk_56      2             37\n",
      "tsk_59      2             35\n",
      "tsk_51      2             35\n",
      "tsk_55      2             28\n",
      "tsk_54      2             28\n",
      "tsk_50      2             24\n",
      "tsk_51      1             22\n",
      "tsk_54      1             15\n",
      "tsk_57      2             14\n",
      "tsk_55      1             11\n",
      "tsk_52      1              8\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "archaic_snp_counts_per_individual_haplotype = df_single_snp.value_counts(['individual', 'haplotype'])\n",
    "print(archaic_snp_counts_per_individual_haplotype)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "e78c7394",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "43.55"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "archaic_snp_counts_per_individual_haplotype.mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e4093cd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "individual  haplotype  segment\n",
      "tsk_56      1          0          74\n",
      "tsk_57      1          0          74\n",
      "tsk_53      2          4          42\n",
      "tsk_52      2          4          38\n",
      "                       5          36\n",
      "tsk_51      2          0          34\n",
      "tsk_50      1          1          33\n",
      "tsk_59      1          4          30\n",
      "tsk_56      2          11         29\n",
      "tsk_57      1          10         28\n",
      "tsk_50      1          10         28\n",
      "tsk_54      2          6          27\n",
      "tsk_58      1          1          27\n",
      "tsk_53      1          6          27\n",
      "tsk_55      2          6          27\n",
      "tsk_59      1          5          24\n",
      "tsk_56      1          3          24\n",
      "tsk_57      1          7          23\n",
      "tsk_59      2          9          22\n",
      "tsk_58      2          10         22\n",
      "tsk_51      1          13         21\n",
      "tsk_50      2          13         21\n",
      "tsk_59      1          2          19\n",
      "tsk_54      1          4          14\n",
      "tsk_57      2          4          12\n",
      "tsk_59      2          2          12\n",
      "tsk_58      1          8          12\n",
      "                       12         11\n",
      "            2          8          10\n",
      "tsk_55      1          4          10\n",
      "tsk_58      2          12          9\n",
      "tsk_53      1          10          9\n",
      "tsk_52      1          10          7\n",
      "tsk_53      1          0           6\n",
      "tsk_56      2          3           6\n",
      "tsk_50      2          1           2\n",
      "tsk_52      2          1           1\n",
      "tsk_59      1          1           1\n",
      "tsk_52      1          1           1\n",
      "tsk_58      2          1           1\n",
      "tsk_59      1          9           1\n",
      "            2          1           1\n",
      "tsk_57      2          1           1\n",
      "tsk_51      1          1           1\n",
      "tsk_57      2          7           1\n",
      "tsk_54      2          1           1\n",
      "tsk_57      1          1           1\n",
      "tsk_50      2          0           1\n",
      "tsk_53      1          1           1\n",
      "tsk_51      2          1           1\n",
      "tsk_56      2          0           1\n",
      "            1          1           1\n",
      "tsk_53      2          1           1\n",
      "tsk_55      2          1           1\n",
      "            1          1           1\n",
      "tsk_54      1          1           1\n",
      "tsk_56      2          1           1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "archaic_snp_counts_per_individual_haplotype_segment = df_single_snp.value_counts(['individual', 'haplotype', 'segment'])\n",
    "print(archaic_snp_counts_per_individual_haplotype_segment)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gaishi-analysis150126",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
