{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dcb26a19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jalh/anaconda3/envs/gaishi-analysis150126/lib/python3.9/site-packages/pandas/core/computation/expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.7.3' currently installed).\n",
      "  from pandas.core.computation.check import NUMEXPR_INSTALLED\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from sprime_mapping import map_sprime_segments\n",
    "from sprime_info import compute_sprime_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "165acd43",
   "metadata": {},
   "outputs": [],
   "source": [
    "# these two files are necessary: score_file with SPrime output and original vcf_file output\n",
    "vcf_file = \"archie.1_biallelic.vcf.gz\"\n",
    "score_file = \"sprime.1src.out.100000.score\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4766a6e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for getting only the introgressed tracts in the target individuals, we also have to provide a list with the individual names (have to correspond to the vcf-file) we are interested in\n",
    "# if not provided, for all individuals in the vcf-file the introgression status is checked (including reference individuals etc. - this is usually not desired / reasonable, but perhaps one want to check)\n",
    "tgt_individuals_file = \"archie.1.tgt.ind.list\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "709e4d40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# if we ant to write the file, also an out_file has to be scpecified\n",
    "#out_file_wo_filtering = \"archie.0.sprime.100000.inferred_wo_filtering.bed\"\n",
    "#out_file_every_snp = \"archie.0.sprime.100000.inferred_every_snp.bed\"\n",
    "#out_file_frac_50 = \"archie.0.sprime.100000.inferred_fraction_050.bed\"\n",
    "out_file_frac_100 = \"archie.0.sprime.100000.inferred_fraction_100.bed\"\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "15c4179f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# open file with tgt individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2a5c5620",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(tgt_individuals_file) as f:\n",
    "    tgt_individuals = [line.strip() for line in f if line.strip()]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "7575004e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['tsk_50',\n",
       " 'tsk_51',\n",
       " 'tsk_52',\n",
       " 'tsk_53',\n",
       " 'tsk_54',\n",
       " 'tsk_55',\n",
       " 'tsk_56',\n",
       " 'tsk_57',\n",
       " 'tsk_58',\n",
       " 'tsk_59']"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tgt_individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "29339831",
   "metadata": {},
   "outputs": [],
   "source": [
    "### Let's also check the sprime output file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c0cdbb4c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    SEGMENT  SNP_COUNT\n",
      "0         0         75\n",
      "1         1         47\n",
      "2         4         42\n",
      "3         5         36\n",
      "4         2         31\n",
      "5         3         30\n",
      "6        11         29\n",
      "7        10         28\n",
      "8         6         27\n",
      "9         7         24\n",
      "10        9         23\n",
      "11        8         22\n",
      "12       13         21\n",
      "13       12         20\n",
      "455\n"
     ]
    }
   ],
   "source": [
    "df = pd.read_csv(score_file, sep=\"\\t\")\n",
    "\n",
    "# Count the number of SNPs per SEGMENT\n",
    "snp_counts = df['SEGMENT'].value_counts()\n",
    "\n",
    "snp_counts_df = snp_counts.reset_index()\n",
    "snp_counts_df.columns = ['SEGMENT', 'SNP_COUNT']\n",
    "\n",
    "print(snp_counts_df)\n",
    "print(snp_counts_df[\"SNP_COUNT\"].sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33b3ec7e",
   "metadata": {},
   "source": [
    "## we assume that only matches of full segments - i.e. each archaic SNP of the segment in the score file has to be present in the target individual - are acceptable\n",
    "\n",
    "## This is done by setting segment_fraction=1, i.e. the full segment without deviations must be matched\n",
    "\n",
    "### With these settings, the first and last SNP, which matches the archaic SNPs of the segment, is used as border of the segment\n",
    "\n",
    "### In this run of the function, we create an output-file (bed-file)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a2ee3d25",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_100 = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    out_file_frac_100, only_tract_output=True, phased=True, return_full_records=False, target_individuals=tgt_individuals,\n",
    "    segment_fraction=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39d0b1c5",
   "metadata": {},
   "source": [
    "### we can also show the full records, including the nsnps in each segment, the fraction of archaic snps in the segment for a specific individual and the snps present\n",
    "### out_file is here set to None, so no new out_file is created"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c5941039",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_100_full = map_sprime_segments(\n",
    "    score_file,\n",
    "    vcf_file,\n",
    "    None, only_tract_output=True, phased=True, return_full_records=True, target_individuals=tgt_individuals,\n",
    "    segment_fraction=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3745e517",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n"
     ]
    }
   ],
   "source": [
    "# now the number of fragments in the dataframe is considerably smaller than for other approachs accepting partial matches\n",
    "print(len(df_100_full))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46c95c01",
   "metadata": {},
   "source": [
    "# We see that - because we require that full segments, i.e. all SNPs of a segment, have to be matched - not all segments are matched"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b8351b83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>all_snps_present</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>8413208</td>\n",
       "      <td>8484292</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>[8413208, 8418944, 8423046, 8430926, 8432462, ...</td>\n",
       "      <td>[8413208, 8418944, 8423046, 8430926, 8432462, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>9931398</td>\n",
       "      <td>10038950</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>9931398</td>\n",
       "      <td>10038950</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>9931398</td>\n",
       "      <td>10038950</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>11390420</td>\n",
       "      <td>11569377</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "      <td>42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>[11390420, 11394987, 11398157, 11407918, 11427...</td>\n",
       "      <td>[11390420, 11394987, 11398157, 11407918, 11427...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>17813068</td>\n",
       "      <td>17870703</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>[17813068, 17815969, 17816939, 17817828, 17819...</td>\n",
       "      <td>[17813068, 17815969, 17816939, 17817828, 17819...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>20647984</td>\n",
       "      <td>20693924</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>20647984</td>\n",
       "      <td>20693924</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end individual  haplotype  nsnps  segment_fraction  \\\n",
       "0      1    238504    314341     tsk_50          1     28               1.0   \n",
       "1      1    238504    314341     tsk_57          1     28               1.0   \n",
       "2      1   8413208   8484292     tsk_56          2     29               1.0   \n",
       "3      1   9931398  10038950     tsk_53          1     27               1.0   \n",
       "4      1   9931398  10038950     tsk_54          2     27               1.0   \n",
       "5      1   9931398  10038950     tsk_55          2     27               1.0   \n",
       "6      1  11390420  11569377     tsk_53          2     42               1.0   \n",
       "7      1  17813068  17870703     tsk_52          2     36               1.0   \n",
       "8      1  20647984  20693924     tsk_50          2     21               1.0   \n",
       "9      1  20647984  20693924     tsk_51          1     21               1.0   \n",
       "\n",
       "   segment                                   all_snps_present  \\\n",
       "0       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "1       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "2       11  [8413208, 8418944, 8423046, 8430926, 8432462, ...   \n",
       "3        6  [9931398, 9939286, 9945995, 9946088, 9951559, ...   \n",
       "4        6  [9931398, 9939286, 9945995, 9946088, 9951559, ...   \n",
       "5        6  [9931398, 9939286, 9945995, 9946088, 9951559, ...   \n",
       "6        4  [11390420, 11394987, 11398157, 11407918, 11427...   \n",
       "7        5  [17813068, 17815969, 17816939, 17817828, 17819...   \n",
       "8       13  [20647984, 20648821, 20649517, 20649859, 20650...   \n",
       "9       13  [20647984, 20648821, 20649517, 20649859, 20650...   \n",
       "\n",
       "                                archaic_snps_present  \n",
       "0  [238504, 251513, 255766, 257368, 257723, 25802...  \n",
       "1  [238504, 251513, 255766, 257368, 257723, 25802...  \n",
       "2  [8413208, 8418944, 8423046, 8430926, 8432462, ...  \n",
       "3  [9931398, 9939286, 9945995, 9946088, 9951559, ...  \n",
       "4  [9931398, 9939286, 9945995, 9946088, 9951559, ...  \n",
       "5  [9931398, 9939286, 9945995, 9946088, 9951559, ...  \n",
       "6  [11390420, 11394987, 11398157, 11407918, 11427...  \n",
       "7  [17813068, 17815969, 17816939, 17817828, 17819...  \n",
       "8  [20647984, 20648821, 20649517, 20649859, 20650...  \n",
       "9  [20647984, 20648821, 20649517, 20649859, 20650...  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_100_full"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce74c73b",
   "metadata": {},
   "source": [
    "# Computation of averages\n",
    "\n",
    "### We now additionally compute the length of the regions as well as the segments per individual\n",
    "## afterwards, we can average over these values and get the desired results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f2d9b91f",
   "metadata": {},
   "outputs": [],
   "source": [
    "res_frac100 = compute_sprime_info(df_100_full)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "19e8e9f1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>total_length</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>57635</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>178957</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>71084</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual_haplotype  total_length individual  haplotype\n",
       "0             tsk_50_1         75837     tsk_50          1\n",
       "1             tsk_50_2         45940     tsk_50          2\n",
       "2             tsk_51_1         45940     tsk_51          1\n",
       "3             tsk_52_2         57635     tsk_52          2\n",
       "4             tsk_53_1        107552     tsk_53          1\n",
       "5             tsk_53_2        178957     tsk_53          2\n",
       "6             tsk_54_2        107552     tsk_54          2\n",
       "7             tsk_55_2        107552     tsk_55          2\n",
       "8             tsk_56_2         71084     tsk_56          2\n",
       "9             tsk_57_1         75837     tsk_57          1"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total length of introgressed regions per individual/haplotype\n",
    "res_frac100[\"total_length_ind_hap\"]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "4dcb5e1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>segment</th>\n",
       "      <th>total_length</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>10</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>13</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>13</td>\n",
       "      <td>45940</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>5</td>\n",
       "      <td>57635</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>6</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>4</td>\n",
       "      <td>178957</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>6</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>6</td>\n",
       "      <td>107552</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>11</td>\n",
       "      <td>71084</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>10</td>\n",
       "      <td>75837</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual_haplotype  segment  total_length individual  haplotype\n",
       "0             tsk_50_1       10         75837     tsk_50          1\n",
       "1             tsk_50_2       13         45940     tsk_50          2\n",
       "2             tsk_51_1       13         45940     tsk_51          1\n",
       "3             tsk_52_2        5         57635     tsk_52          2\n",
       "4             tsk_53_1        6        107552     tsk_53          1\n",
       "5             tsk_53_2        4        178957     tsk_53          2\n",
       "6             tsk_54_2        6        107552     tsk_54          2\n",
       "7             tsk_55_2        6        107552     tsk_55          2\n",
       "8             tsk_56_2       11         71084     tsk_56          2\n",
       "9             tsk_57_1       10         75837     tsk_57          1"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# total length of introgressed regions per individual/haplotype and segment\n",
    "res_frac100[\"total_length_ind_hap_seg\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "bf8a56e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>segments</th>\n",
       "      <th>nr_segments</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>[13]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>[13]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>[5]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>[6]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>[4]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>[6]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>[6]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>[11]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>[10]</td>\n",
       "      <td>1</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual_haplotype segments  nr_segments individual  haplotype\n",
       "0             tsk_50_1     [10]            1     tsk_50          1\n",
       "1             tsk_50_2     [13]            1     tsk_50          2\n",
       "2             tsk_51_1     [13]            1     tsk_51          1\n",
       "3             tsk_52_2      [5]            1     tsk_52          2\n",
       "4             tsk_53_1      [6]            1     tsk_53          1\n",
       "5             tsk_53_2      [4]            1     tsk_53          2\n",
       "6             tsk_54_2      [6]            1     tsk_54          2\n",
       "7             tsk_55_2      [6]            1     tsk_55          2\n",
       "8             tsk_56_2     [11]            1     tsk_56          2\n",
       "9             tsk_57_1     [10]            1     tsk_57          1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#segments per individual/haplotype \n",
    "res_frac100[\"segments_per_ind\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "914f14bb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>nsnps</th>\n",
       "      <th>segment_fraction</th>\n",
       "      <th>segment</th>\n",
       "      <th>all_snps_present</th>\n",
       "      <th>archaic_snps_present</th>\n",
       "      <th>individual_haplotype</th>\n",
       "      <th>interval_length</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>tsk_50_1</td>\n",
       "      <td>75837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>238504</td>\n",
       "      <td>314341</td>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "      <td>28</td>\n",
       "      <td>1.0</td>\n",
       "      <td>10</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>[238504, 251513, 255766, 257368, 257723, 25802...</td>\n",
       "      <td>tsk_57_1</td>\n",
       "      <td>75837</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>8413208</td>\n",
       "      <td>8484292</td>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>29</td>\n",
       "      <td>1.0</td>\n",
       "      <td>11</td>\n",
       "      <td>[8413208, 8418944, 8423046, 8430926, 8432462, ...</td>\n",
       "      <td>[8413208, 8418944, 8423046, 8430926, 8432462, ...</td>\n",
       "      <td>tsk_56_2</td>\n",
       "      <td>71084</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>9931398</td>\n",
       "      <td>10038950</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>tsk_53_1</td>\n",
       "      <td>107552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>9931398</td>\n",
       "      <td>10038950</td>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>tsk_54_2</td>\n",
       "      <td>107552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>9931398</td>\n",
       "      <td>10038950</td>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "      <td>27</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>[9931398, 9939286, 9945995, 9946088, 9951559, ...</td>\n",
       "      <td>tsk_55_2</td>\n",
       "      <td>107552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>11390420</td>\n",
       "      <td>11569377</td>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "      <td>42</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "      <td>[11390420, 11394987, 11398157, 11407918, 11427...</td>\n",
       "      <td>[11390420, 11394987, 11398157, 11407918, 11427...</td>\n",
       "      <td>tsk_53_2</td>\n",
       "      <td>178957</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>17813068</td>\n",
       "      <td>17870703</td>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5</td>\n",
       "      <td>[17813068, 17815969, 17816939, 17817828, 17819...</td>\n",
       "      <td>[17813068, 17815969, 17816939, 17817828, 17819...</td>\n",
       "      <td>tsk_52_2</td>\n",
       "      <td>57635</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>20647984</td>\n",
       "      <td>20693924</td>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>tsk_50_2</td>\n",
       "      <td>45940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>20647984</td>\n",
       "      <td>20693924</td>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "      <td>21</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>[20647984, 20648821, 20649517, 20649859, 20650...</td>\n",
       "      <td>tsk_51_1</td>\n",
       "      <td>45940</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chrom     start       end individual  haplotype  nsnps  segment_fraction  \\\n",
       "0      1    238504    314341     tsk_50          1     28               1.0   \n",
       "1      1    238504    314341     tsk_57          1     28               1.0   \n",
       "2      1   8413208   8484292     tsk_56          2     29               1.0   \n",
       "3      1   9931398  10038950     tsk_53          1     27               1.0   \n",
       "4      1   9931398  10038950     tsk_54          2     27               1.0   \n",
       "5      1   9931398  10038950     tsk_55          2     27               1.0   \n",
       "6      1  11390420  11569377     tsk_53          2     42               1.0   \n",
       "7      1  17813068  17870703     tsk_52          2     36               1.0   \n",
       "8      1  20647984  20693924     tsk_50          2     21               1.0   \n",
       "9      1  20647984  20693924     tsk_51          1     21               1.0   \n",
       "\n",
       "   segment                                   all_snps_present  \\\n",
       "0       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "1       10  [238504, 251513, 255766, 257368, 257723, 25802...   \n",
       "2       11  [8413208, 8418944, 8423046, 8430926, 8432462, ...   \n",
       "3        6  [9931398, 9939286, 9945995, 9946088, 9951559, ...   \n",
       "4        6  [9931398, 9939286, 9945995, 9946088, 9951559, ...   \n",
       "5        6  [9931398, 9939286, 9945995, 9946088, 9951559, ...   \n",
       "6        4  [11390420, 11394987, 11398157, 11407918, 11427...   \n",
       "7        5  [17813068, 17815969, 17816939, 17817828, 17819...   \n",
       "8       13  [20647984, 20648821, 20649517, 20649859, 20650...   \n",
       "9       13  [20647984, 20648821, 20649517, 20649859, 20650...   \n",
       "\n",
       "                                archaic_snps_present individual_haplotype  \\\n",
       "0  [238504, 251513, 255766, 257368, 257723, 25802...             tsk_50_1   \n",
       "1  [238504, 251513, 255766, 257368, 257723, 25802...             tsk_57_1   \n",
       "2  [8413208, 8418944, 8423046, 8430926, 8432462, ...             tsk_56_2   \n",
       "3  [9931398, 9939286, 9945995, 9946088, 9951559, ...             tsk_53_1   \n",
       "4  [9931398, 9939286, 9945995, 9946088, 9951559, ...             tsk_54_2   \n",
       "5  [9931398, 9939286, 9945995, 9946088, 9951559, ...             tsk_55_2   \n",
       "6  [11390420, 11394987, 11398157, 11407918, 11427...             tsk_53_2   \n",
       "7  [17813068, 17815969, 17816939, 17817828, 17819...             tsk_52_2   \n",
       "8  [20647984, 20648821, 20649517, 20649859, 20650...             tsk_50_2   \n",
       "9  [20647984, 20648821, 20649517, 20649859, 20650...             tsk_51_1   \n",
       "\n",
       "   interval_length  \n",
       "0            75837  \n",
       "1            75837  \n",
       "2            71084  \n",
       "3           107552  \n",
       "4           107552  \n",
       "5           107552  \n",
       "6           178957  \n",
       "7            57635  \n",
       "8            45940  \n",
       "9            45940  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# the df now also has interval length\n",
    "res_frac100[\"df\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0cbd33d1",
   "metadata": {},
   "source": [
    "## Average number of segments per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a09eed12",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum_per_individual = (\n",
    "    res_frac100[\"segments_per_ind\"].groupby(\"individual\")[\"nr_segments\"]\n",
    "      .sum()\n",
    "      .reset_index(name=\"sum_nr_segments\")\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "21a6c628",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>sum_nr_segments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_52</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_54</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_55</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_56</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  sum_nr_segments\n",
       "0     tsk_50                2\n",
       "1     tsk_51                1\n",
       "2     tsk_52                1\n",
       "3     tsk_53                2\n",
       "4     tsk_54                1\n",
       "5     tsk_55                1\n",
       "6     tsk_56                1\n",
       "7     tsk_57                1"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_per_individual\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5d6b4fcb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.25\n"
     ]
    }
   ],
   "source": [
    "avg_segments_per_individual = sum_per_individual[\"sum_nr_segments\"].mean()\n",
    "print(avg_segments_per_individual)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34977346",
   "metadata": {},
   "source": [
    "## or per chromosome, ,i.e. individual haplotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "a681f619",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum_per_individual_and_haplotype = (\n",
    "    res_frac100[\"segments_per_ind\"].groupby([\"individual\", \"haplotype\"])[\"nr_segments\"]\n",
    "      .sum()\n",
    "      .reset_index(name=\"sum_nr_segments\")\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "73f12fa2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>individual</th>\n",
       "      <th>haplotype</th>\n",
       "      <th>sum_nr_segments</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>tsk_50</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>tsk_50</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>tsk_51</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>tsk_52</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>tsk_53</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>tsk_53</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>tsk_54</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>tsk_55</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>tsk_56</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>tsk_57</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  individual  haplotype  sum_nr_segments\n",
       "0     tsk_50          1                1\n",
       "1     tsk_50          2                1\n",
       "2     tsk_51          1                1\n",
       "3     tsk_52          2                1\n",
       "4     tsk_53          1                1\n",
       "5     tsk_53          2                1\n",
       "6     tsk_54          2                1\n",
       "7     tsk_55          2                1\n",
       "8     tsk_56          2                1\n",
       "9     tsk_57          1                1"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum_per_individual_and_haplotype"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "bab11cc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0\n"
     ]
    }
   ],
   "source": [
    "avg_segments_per_individual_haplotype = sum_per_individual_and_haplotype[\"sum_nr_segments\"].mean()\n",
    "print(avg_segments_per_individual_haplotype)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "gaishi-analysis150126",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.19"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
